File: plumber/gateway/endpoints/setBot.R
#* @param apiKey The API key for authentication
#* @param protocol The protocol to use
#* @param pool The pool to target
#* @param network The network to use
#* @param pair The trading pair
#* @param side The side to set (long, short, hold, or neutral)
#* @param threshold The threshold for the strategy (default is 1)
#* @param max_usd The maximum USD amount (default is 10,000,000)
#* @param slippage The slippage percentage (default is 1)
#* @param share The share percentage (default is 100)
#* @param platform The platform to use (default is uniswapV3)
#* @response 200 Returns the result of setting the bot strategy
#* @response 400 Bad request
#* @response 500 Internal server error
#* @tag managers
#* @post /setBot
setBotHandler = function(apiKey,
                         protocol = "dhedge",
                         pool,
                         network,
                         pair,
                         side,
                         threshold = 1,
                         max_usd = 10000000,
                         slippage = 1,
                         share = 100,
                         platform = "odos",
			 lending=FALSE) {
  
  # Normalize case
  protocol <- tolower(protocol)
  pool <- tolower(pool)
  network <- tolower(network)
  side <- tolower(side)
  platform <- tolower(platform)
  if (is.null(lending) || !is.logical(lending)) { lending = FALSE }
  # API key check
  check <- basic_check(network = network, protocol = protocol, pool = pool, apiKey = apiKey)
  if (check$status == "fail") return(check)
  
  # Convert numeric inputs
  if (!is.null(max_usd)) {
    max_usd <- suppressWarnings(as.numeric(max_usd))
  }
  slippage <- suppressWarnings(as.numeric(slippage))
  share <- suppressWarnings(as.numeric(share))
  threshold <- suppressWarnings(as.numeric(threshold))
  # Validate 'side'
  if (!(side %in% c("hold", "neutral", "short", "long"))) {
    return(list(status = "fail", status_code = 1008, message = "The specified side must be one of: long, short, hold, or neutral."))
  }
  else if (side == "short")  {
	 return(list(status = "fail", status_code = 400, message ="Shorting is temporary disabled."))
	 if (network != "arbitrum" || network != "optimism") return(list(status = "fail", status_code = 400, message = "Shorts doesn't work on the specified network, use Arbitrum or Optimism to go short."))
  }
  # Validate 'threshold'
  if (!is.na(threshold)) {
    if (threshold >= 0 && threshold <= 100) {
      threshold <- round(threshold,2)
    } else {
      return(list(status = "fail", status_code = 400, message = "Threshold must be number in the range [0, 100]."))
    }
  } else {
    return(list(status = "fail", status_code = 400, message = "Threshold is not a valid number in the range [0, 100]."))
  }
  
  # Validate 'share'
  if (!is.na(share)) {
    if (share >= 1 && share <= 100) {
      share <- round(share,2)
    } else {
      return(list(status = "fail", status_code = 400, message = "Share must be a number in the range [1, 100]."))
    }
  } else {
    return(list(status = "fail", status_code = 400, message = "Share is not a valid number in the range [1, 100]."))
  }
  
  # Validate 'max_usd'
  if (!is.null(max_usd)) {
    if (is.na(max_usd)) {
      return(list(status = "fail", status_code = 400, message = "The specified max_usd is not numeric."))
    }
    if (max_usd <= 0) {
      return(list(status = "fail", status_code = 400, message = "The specified max_usd must be a number > 0."))
    }
    max_usd <- round(max_usd, 2)
  }
  
  # Build URL
  url <- paste0(pep, "setSide?",
                "apiKey=", apiKey,
                "&protocol=", protocol,
                "&pool=", pool,
                "&network=", network,
                "&pair=", pair,
                "&side=", side,
                "&threshold=", threshold,
                "&max_usd=", max_usd,
                "&slippage=", slippage,
                "&share=", share,
                "&platform=", platform,
  		"&lending=",lending	
  	)
  
  # Masked API key for logging
  masked_api <- mask_api(apiKey)
  
  # Perform POST request to /setSide
  response <- POST(url)
  content_response <- content(response, "text")
  
  parsed_response <- tryCatch({
    fromJSON(content_response)
  }, error = function(e) {
    list(status = "fail", status_code = 500, message = paste0("Error parsing response: ", content_response))
  })
  
  # Log message
  msg <- paste0(
    "setBot invoked | apiKey: ", masked_api,
    " / pool: ", pool,
    " / protocol: ", protocol,
    " / network: ", network,
    " / pair: ", pair,
    " / side: ", side,
    " / threshold: ", threshold,
    " / max_usd: ", max_usd,
    " / slippage: ", slippage,
    " / share: ", share,
    " / platform: ", platform,
    " / lending: ",lending,
    " / response: ", paste0(names(parsed_response), "=", unlist(parsed_response), collapse = ", ")
  )
  
  print(msg)
  #discord(msg = msg, channel = "#api-logs")
  send_telegram_text(msg)
  return(parsed_response)
}

pr$handle("POST","/setBot",setBotHandler, comment="This endpoint is used to set the sides of your tradingbot strategy on a specific pool, network and protocol. The sides are used by the tradebots to monitor and rebalance the pools according to the strategy. The sides can be long, short, hold or neutral. Long will buy everything on the vault, short will short everything, neutral will sell to USDC all positions and hold wont do anything with new deposits or current positions. Enable lending to allow the bot to deposit the token on the current side into AAVE V3 to earn yield while you hold. (Require AAVE V3 Enabled, Coming Soon)")

File: infinitetrading_api/express/src/requests/setBot.ts
/**
 * requests/setBot.ts — Node port of R gateway's setBot.R (proxies to
 * plumber-api port 8002's setSideHandler in src/api/api.R, which itself
 * writes to `dhedge_sides` and then invokes executeTrades()).
 *
 * ***HIGH RISK — LIVE TRADING.*** This endpoint both persists the bot's
 * strategy configuration AND immediately triggers a live rebalance via the
 * full tradebot/executeTrades decision engine (see ../tradeEngine.ts).
 *
 * PARITY NOTES:
 *  - Gateway layer (setBot.R):
 *      1. Normalize case (protocol/pool/network/side/platform lowercased).
 *      2. lending defaults to FALSE if null/non-logical.
 *      3. basic_check(network, protocol, pool, apiKey) -- fail -> returned.
 *      4. side must be one of hold/neutral/short/long (else 1008).
 *      5. side=="short" requires network in [arbitrum, optimism] (else 400).
 *      6. threshold must be a number in [0,100] (else 400), rounded to 2dp.
 *      7. share must be a number in [1,100] (else 400), rounded to 2dp.
 *      8. max_usd (if provided) must be numeric > 0 (else 400), rounded 2dp.
 *      9. Proxies to inner /setSide with all params + lending.
 *  - Inner (port 8002) layer (setSideHandler in api.R):
 *      1. isValidApiKey(network, protocol, pool, apiKey) DB check (gas
 *         wallet linked to pool) -- fail -> 401 "Invalid API key".
 *      2. side=="short" re-checked against short_networks (redundant with
 *         gateway check, kept for parity -- defi.R's short_networks list is
 *         identical to the gateway's).
 *      3. setSide(...) -- upserts the dhedge_sides row (db.R's setSide()).
 *      4. executeTrades(...) -- fire-and-forget-ish (result is logged, not
 *         propagated to the response) trade execution based on the new
 *         side/threshold/share/max_usd/platform.
 *      5. Returns the result of setSide() (NOT executeTrades()'s result) --
 *         i.e. the HTTP response reflects "was the DB write successful",
 *         while the actual trade execution result is only logged
 *         server-side. This is intentional/existing behavior (executeTrades
 *         can legitimately fail without failing the overall /setBot call,
 *         e.g. "hold" side, or a trade that doesn't meet threshold), so it
 *         is preserved here exactly.
 */

import { Router, Request, Response } from 'express';
import { dbExecute } from '../db';
import { basicCheck, toRWireFormat } from '../basicCheck';
import { executeTrades } from '../tradeEngine';
import { DEFAULT_PLATFORM } from '../utils/parseDapp';

const router = Router();

const SHORT_NETWORKS = ['arbitrum', 'optimism'];

async function isValidApiKeyForPool(network: string, protocol: string, pool: string, apiKey: string): Promise<boolean> {
  const { dbQuery } = await import('../db');
  try {
    const rows = await dbQuery(
      'SELECT 1 FROM gas_wallets WHERE network = ? AND protocol = ? AND pool = ? AND token = ? LIMIT 1',
      [network, protocol, pool, apiKey]
    );
    return rows.length > 0;
  } catch (e: any) {
    console.log('Error: isValidApiKeyForPool:', e.message);
    return false;
  }
}

// setSide (db.R lines 232-238): upserts dhedge_sides.
async function setSide(opts: {
  network: string; pool: string; pair: string; side: string; threshold: number;
  maxUsd: number; share: number; platform: string; slippage: number;
}): Promise<{ status: string; status_code: number; message: string }> {
  try {
    const pair = opts.pair.toUpperCase();
    await dbExecute(
      `INSERT INTO dhedge_sides (network, pool, pair, side, threshold, max_usd, share, platform, slippage)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
       ON DUPLICATE KEY UPDATE side=VALUES(side), threshold=VALUES(threshold),
       max_usd=VALUES(max_usd), share=VALUES(share), platform=VALUES(platform), slippage=VALUES(slippage)`,
      [opts.network, opts.pool, pair, opts.side, opts.threshold, Math.trunc(opts.maxUsd), opts.share, opts.platform, opts.slippage]
    );
    return { status: 'success', status_code: 200, message: 'Sides submitted successfully' };
  } catch (e: any) {
    console.log(`Error: setSide for network: ${opts.network} pool: ${opts.pool} error: ${e.message}`);
    return { status: 'fail', status_code: 500, message: 'Internal error submitting sides' };
  }
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

/**
 * @openapi
 * /setBot:
 *   post:
 *     summary: Set the trading-bot strategy (side/threshold/share) for a pool and trigger a rebalance (HIGH RISK, live trading)
 *     tags: [Managers]
 *     parameters:
 *       - in: query
 *         name: apiKey
 *         required: true
 *         schema: { type: string }
 *       - in: query
 *         name: protocol
 *         schema: { type: string, enum: [dhedge, chamber, defund], default: dhedge }
 *       - in: query
 *         name: pool
 *         required: true
 *         schema: { type: string }
 *       - in: query
 *         name: network
 *         required: true
 *         schema: { type: string, enum: [base, optimism, arbitrum, polygon, ethereum, mainnet, hyperliquid] }
 *       - in: query
 *         name: pair
 *         required: true
 *         schema: { type: string }
 *       - in: query
 *         name: side
 *         required: true
 *         schema: { type: string, enum: [long, short, hold, neutral] }
 *       - in: query
 *         name: threshold
 *         schema: { type: number, default: 1 }
 *       - in: query
 *         name: max_usd
 *         schema: { type: number, default: 10000000 }
 *       - in: query
 *         name: slippage
 *         schema: { type: number, default: 1 }
 *       - in: query
 *         name: share
 *         schema: { type: number, default: 100 }
 *       - in: query
 *         name: platform
 *         schema: { type: string, enum: [auto, uniswapV3, velodrome, velodromecl, aerodrome, aerodromecl, pancakecl, quickswap, kyberswap, cowswap, pendle, aavev3, compoundv3, fluid, lyra, hyperliquid], default: auto }
 *       - in: query
 *         name: lending
 *         schema: { type: boolean, default: false }
 *     responses:
 *       200:
 *         description: Sides submitted successfully (trade execution result is logged server-side only).
 */
router.post('/setBot', async (req: Request, res: Response) => {
  const q = { ...req.query, ...req.body };
  const apiKey = String(q.apiKey || '');
  const protocol = String(q.protocol || 'dhedge').toLowerCase();
  const pool = String(q.pool || '').toLowerCase();
  const network = String(q.network || '').toLowerCase();
  const pair = String(q.pair || '');
  const side = String(q.side || '').toLowerCase();
  const platform = String(q.platform || DEFAULT_PLATFORM).toLowerCase();
  let lending = false;
  if (typeof q.lending === 'boolean') lending = q.lending;
  else if (typeof q.lending === 'string') lending = q.lending.toLowerCase() === 'true';

  // === Gateway layer: basicCheck ===
  const check = await basicCheck({ network, protocol, pool, apiKey });
  if (check.status === 'fail') {
    return res.json(toRWireFormat(check));
  }

  const maxUsdRaw = q.max_usd === undefined ? '10000000' : String(q.max_usd);
  const slippageRaw = q.slippage === undefined ? '1' : String(q.slippage);
  const shareRaw = q.share === undefined ? '100' : String(q.share);
  const thresholdRaw = q.threshold === undefined ? '1' : String(q.threshold);

  const maxUsdNum = Number(maxUsdRaw);
  const slippageNum = Number(slippageRaw);
  const shareNum = Number(shareRaw);
  const thresholdNum = Number(thresholdRaw);

  // === Gateway layer: side validation ===
  if (!['hold', 'neutral', 'short', 'long'].includes(side)) {
    return res.json({ status: ['fail'], status_code: [1008], message: ['The specified side must be one of: long, short, hold, or neutral.'] });
  }
  if (side === 'short' && !SHORT_NETWORKS.includes(network)) {
    return res.json({ status: ['fail'], status_code: [400], message: ['Shorts only supported on Arbitrum or Optimism.'] });
  }

  // === Gateway layer: threshold validation ===
  if (Number.isNaN(thresholdNum) || thresholdNum < 0 || thresholdNum > 100) {
    return res.json({ status: ['fail'], status_code: [400], message: ['Threshold must be number in the range [0, 100].'] });
  }
  const threshold = round2(thresholdNum);

  // === Gateway layer: share validation ===
  if (Number.isNaN(shareNum) || shareNum < 1 || shareNum > 100) {
    return res.json({ status: ['fail'], status_code: [400], message: ['Share must be a number in the range [1, 100].'] });
  }
  const share = round2(shareNum);

  // === Gateway layer: max_usd validation ===
  if (Number.isNaN(maxUsdNum)) {
    return res.json({ status: ['fail'], status_code: [400], message: ['The specified max_usd is not numeric.'] });
  }
  if (maxUsdNum <= 0) {
    return res.json({ status: ['fail'], status_code: [400], message: ['The specified max_usd must be a number > 0.'] });
  }
  const maxUsd = round2(maxUsdNum);

  // === Inner layer: isValidApiKey DB check ===
  const validForPool = await isValidApiKeyForPool(network, protocol, pool, apiKey);
  if (!validForPool) {
    return res.json({ status: ['fail'], status_code: [401], message: ['Invalid API key'] });
  }

  // === Inner layer: re-check short_networks (parity with R's redundant check) ===
  if (side === 'short' && !SHORT_NETWORKS.includes(network)) {
    return res.json({ status: ['fail'], status_code: [401], message: ['Shorting is not allowed on the specified network'] });
  }

  // === Inner layer: setSide DB upsert ===
  const setSideResult = await setSide({ network, pool, pair, side, threshold, maxUsd, share, platform, slippage: slippageNum });
  console.log(`/setSide: ${JSON.stringify(setSideResult)}`);

  // === Inner layer: executeTrades (result logged only, not returned) ===
  executeTrades({
    pool, pair, side: side as any, share, threshold, slippage: slippageNum,
    apiKey, maxUsd, composition: null, platform, protocol, network,
  }).then((executeTradesRes) => {
    console.log(`/setSide: executeTrades response: ${JSON.stringify(executeTradesRes)}`);
  }).catch((e) => {
    console.log(`/setSide: executeTrades error: ${e.message}`);
  });

  const maskedApi = apiKey.length > 8 ? `${apiKey.slice(0, 4)}...${apiKey.slice(-4)}` : '****';
  console.log(`setBot invoked | apiKey: ${maskedApi} / pool: ${pool} / protocol: ${protocol} / network: ${network} / pair: ${pair} / side: ${side} / threshold: ${threshold} / max_usd: ${maxUsd} / slippage: ${slippageNum} / share: ${share} / platform: ${platform} / lending: ${lending}`);

  return res.json({ status: [setSideResult.status], status_code: [setSideResult.status_code], message: [setSideResult.message] });
});

export default router;
